### PR TITLE
[WIP] Fix comments in template literals

### PR DESCRIPTION
**Pull request by yamafaktory**
I was experimenting with #623. Not quite sure how the comments should be printed in this case...

@vjeux the example you provided now ends up being printed as follow:

```js
`
${ExampleStory.getFragment("story")} // $FlowFixMe found when converting React.createClass to ES6
`;
```

### CODE REVIEW

**Comment by yamafaktory** [comment]
Oh, wait this is completely wrong!

**Comment by yamafaktory** [comment]
Now this is much better (and valid...):

```js
`
${// $FlowFixMe found when converting React.createClass to ES6
ExampleStory.getFragment("story")}
`;
```

But there's still another edge case that needs to be handled (not from the initial issue):

```js
`
${// $FlowFixMe found when converting React.createClass to ES6
ExampleStory.getFragment('story')}
${a/*asdassd*/}
```

Turns into:

```js
`
${// $FlowFixMe found when converting React.createClass to ES6
ExampleStory.getFragment("story")}
${a}/*asdassd*/
`;
```

**Comment by yamafaktory** [comment]
Now this is able to handle something like:

```js
`
${// $FlowFixMe found when converting React.createClass to ES6
ExampleStory.getFragment('story')
// Another comment
}
`
```

Which turns into:

```js
`
${// $FlowFixMe found when converting React.createClass to ES6
ExampleStory.getFragment("story")
/* Another comment*/}
`;
```

I will have to investigate why the trailing comment is turned into a block comment by the way...

**Comment by yamafaktory** [comment]
@vjeux actually the trailing comment is transformed into a block by Flow. Without passing the `--parser flow` flag everything is fine.

**Comment by jlongster** [comment]
I need to switch to other work right now, sorry I didn't get to this yet :) I will try to look later tonight, thanks!

File: src/comments.js [hunk 1]
```diff
@@ -144,7 +144,8 @@ function attach(comments, ast, text) {
       if (
         handleMemberExpressionComment(enclosingNode, followingNode, comment) ||
           handleIfStatementComments(enclosingNode, followingNode, comment) ||
-          handleTryStatementComments(enclosingNode, followingNode, comment)
+          handleTryStatementComments(enclosingNode, followingNode, comment, true) ||
```
Comment (vjeux) [inline src/comments.js]: Why do you change TryStatement?

File: src/comments.js [hunk 2]
```diff
@@ -369,6 +372,22 @@ function handleMemberExpressionComment(enclosingNode, followingNode, comment) {
   return false;
 }
 
+function handleTemplateLiteralComment(enclosingNode, comment, isOwnLine) {
+  if (
+    enclosingNode &&
+      enclosingNode.type === "TemplateLiteral"
+  ) {
+    if (isOwnLine) {
+      addTrailingComment(enclosingNode.expressions[0], comment);
```
Comment (vjeux) [inline src/comments.js]: what happens if there's more than one expression? Is it going to pick the right one?

File: src/comments.js [hunk 3]
```diff
@@ -144,7 +144,8 @@ function attach(comments, ast, text) {
       if (
         handleMemberExpressionComment(enclosingNode, followingNode, comment) ||
           handleIfStatementComments(enclosingNode, followingNode, comment) ||
-          handleTryStatementComments(enclosingNode, followingNode, comment)
+          handleTryStatementComments(enclosingNode, followingNode, comment, true) ||
+          handleTemplateLiteralComment(enclosingNode, comment, true)
```
Comment (vjeux) [inline src/comments.js]: Looks like I'm not consistent in adding a s or not at the end of `Comment`, we should normalize this :)

File: src/comments.js (hunk 1 above)
Comment (yamafaktory) [inline src/comments.js]: Oh... it was part of a test... will be cleaned-up!

File: src/comments.js (hunk 3 above)
Comment (yamafaktory) [inline src/comments.js]: Ok, will fix it that!

**Comment by yamafaktory** [comment]
@vjeux I've implemented a new logic for handling the comments when more than one expression is present in the template literal.

**Comment by yamafaktory** [comment]
@vjeux Can you please review it again?

**Comment by yamafaktory** [comment]
@vjeux your opinion on that one after refactoring?

**Comment by vjeux** [comment]
I'm sorry, I looked at this multiple times but I don't quite understand how it works. Could you explain to me what the idea is?

File: src/comments.js [hunk 4]
```diff
@@ -425,6 +445,35 @@ function printComment(commentPath) {
   }
 }
 
+function findExpressionIndexForComment (expressions, comment, options) {
+  var match;
+  const startPos = locStart(comment) - 1;
+  const endPos = locEnd(comment) + 1;
+
+  for (var i = 0; i < expressions.length; ++i) {
```
Comment (vjeux) [inline src/comments.js]: Can you make a function

```js
getExpressionRange(expr) {
  if (expr.start) {
    // Babylon
    return {start: expr.start, end: expr.end}
  }
  // Flow
  return {start: expr.range[0], end: expr.range[1]}
}
```

this way you don't have to thread the parser everywhere.

File: src/comments.js [hunk 5]
```diff
@@ -409,6 +411,24 @@ function handleObjectProperty(enclosingNode, precedingNode, comment) {
   return false;
 }
 
+function handleTemplateLiteralComments(enclosingNode, comment, options) {
+  if (
+    enclosingNode &&
+      enclosingNode.type === "TemplateLiteral"
+  ) {
+    const expressionIndex = findExpressionIndexForComment(
```
Comment (yamafaktory) [inline src/comments.js]: @vjeux Sure! First thing, we need to find the correct expression in the template literal to attach the comment.

File: src/comments.js [hunk 6]
```diff
@@ -425,6 +445,35 @@ function printComment(commentPath) {
   }
 }
 
+function findExpressionIndexForComment (expressions, comment, options) {
```
Comment (yamafaktory) [inline src/comments.js]: @vjeux Then this helper function is invoked in order to detect the correct expression's index. Flow and Babylon don't handle it the same way, that's why it's a bit redundant (but still different).

File: src/comments.js (hunk 4 above)
Comment (yamafaktory) [inline src/comments.js]: No pbl, I'll do it 👍.

**Comment by vjeux** [comment]
Okay, so now I understand what you are doing, but unfortunately I don't think that this is going to work.

```js
`${
a +  // a
  a
}`
```

is transformed into

```js
`${a + a}`; // a
```

but the enclosing node of the comment is BinaryExpression and not TemplateLiteral.

This is going to be even more frequent with #749 where we strip all the empty lines.

The more I think about it, the more I think that the correct way to solve this problem is to add the concept of a boundary for `line-suffix` abstraction.

If you reach the end of the `}`, you need to flush the comment right away, you cannot go further.

So maybe what we need is to introduce `line-suffix-boundary` node that does this. What do you think @jlongster

**Comment by vjeux** [comment]
Here's a shot at implementing the idea I just gave: #750

**Comment by jlongster** [comment]
I haven't though about this enough to know if there's a better solution, but that seems like a valid solution. If we find a different way later it's not hard to remove.

Would it be possible to always convert comments into block comments instead?

Or can we convert these special kinds of comments into leading comments, which I think would avoid the whole line suffix thing and add the hardline itself?

I'm not too opinionated about it, just thinking of ideas. The `lineSuffixBoundary` looks OK to me as well.

**Comment by vjeux** [comment]
> Would it be possible to always convert comments into block comments instead?
> Or can we convert these special kinds of comments into leading comments

This is tricky, the code that attaches comments only have access to the enclosing parent, but we need to traverse up the tree in order to know if we are inside of a template literal or jsx expression. We can add support for it but that's going to require some refactoring.

**Comment by jlongster** [comment]
I was wondering if there was a good place to do it post-attachment phase, but if not, that's ok. We can do your line suffix boundary idea if it works!

**Comment by yamafaktory** [comment]
@jlongster & @vjeux I refactored it a bit now, taking into account @jlongster suggestions.

Example:

```js
`
${a // comment
}

${b /* comment */}

${/*comment*/ c /* comment */}

${// comment
d //comment
}
`
```

Is now turned (Flow & Babylon doing the same) into:

```js
`
${/* comment*/
a}

${/* comment */ b}

${/*comment*/ /* comment */ c}

${/* comment*/
/*comment*/
d}
`;
```

Every comment is always turned into a leading block comment.

**Comment by yamafaktory** [comment]
@jlongster any reason why the CI was not triggered?

**Comment by vjeux** [comment]
Okay, we actually need both your fix and mine. I'll go ahead and merge this one and rebase mine on-top. Thanks a lot!

**Comment by yamafaktory** [comment]
@vjeux Thanks for your guidance 👍.

**Pull request merged**